### PR TITLE
Setting up options for ablation studies

### PR DESCRIPTION
**Pull request by lhallee**
Want to be able to toggle between

- Flex attention and SDPA
- AdamW optimizer, Muon, and Muon-Adam ensemble
- Regular transformer vs. Unet + Value embeddings
- Learning rate tuning

### CODE REVIEW

**Comment by lapp0** [comment]
Is there a reason to support `F.scaled_dot_product_attention`? Much simpler to just support cleaner, faster `flex_attention` IMO.

For hyperparameter tuning in general, you might consider making `train` a function parameterized by the `ModelConfig` and `Hyperparameters`. This allows us to easily plug the function into an arbitrary hyperparameter search utility.

You can also easily make the dataclass defined by kwargs, e.g. for modded-gpt I have

```
default_dataclass_map = {"train": Hyperparameters, "gpt": GPTConfig}
def parse_args(dataclass_map=None):
    parser = argparse.ArgumentParser()
    dataclass_map = dataclass_map or default_dataclass_map

    def resolve_type(field):
        origin = get_origin(field.type)
        if origin is Union:
            args = get_args(field.type)
            non_none_types = [arg for arg in args if arg is not type(None)]  # Exclude NoneType
            if len(non_none_types) == 1:
                return non_none_types[0]
        return field.type

    # Dynamically add arguments for each dataclass
    for prefix, dataclass_type in dataclass_map.items():
        for field in fields(dataclass_type):
            arg_name = f"--{prefix}.{field.name}"
            arg_type = resolve_type(field)
            if field.default != MISSING:
                default = field.default
            elif field.default_factory != MISSING:  # Handle default_factory
                default = field.default_factory()
            else:
                default = None
            parser.add_argument(
                arg_name,
                type=arg_type,
                default=default,
                help=f"{field.name} for {prefix} (type: {arg_type.__name__})"
            )
    args = parser.parse_args()
    result = {}
    for prefix, dataclass_type in dataclass_map.items():
        kwargs = {
            field.name: getattr(args, f"{prefix}.{field.name}")
            for field in fields(dataclass_type)
        }
        result[prefix] = dataclass_type(**kwargs)
    return result


if __name__ == "__main__":
    cl_args = parse_args()
    train(gpt_config=cl_args["gpt"], args=cl_args["train"])
```

**Comment by lhallee** [comment]
Going to close this for now and add hyperparam search later
